%% LyX 1.3 created this file.  For more info, see http://www.lyx.org/.
%% Do not edit unless you really know what you are doing.
\documentclass[english, 12pt]{article}
\usepackage{times}
%\usepackage{algorithm2e}
\usepackage{url}
\usepackage{bbm}
\usepackage[T1]{fontenc}
\usepackage[latin1]{inputenc}
\usepackage{geometry}
\geometry{verbose,letterpaper,tmargin=2cm,bmargin=2cm,lmargin=1.5cm,rmargin=1.5cm}
\usepackage{rotating}
\usepackage{color}
\usepackage{graphicx}
\usepackage{amsmath, amsthm, amssymb}
\usepackage{setspace}
\usepackage{lineno}
\usepackage{hyperref}
\usepackage{bbm}
\usepackage{makecell}
\usepackage{placeins}
\usepackage{subcaption}

%\renewcommand{\arraystretch}{1.8}

%\linenumbers
%\doublespacing
\onehalfspacing
%\usepackage[authoryear]{natbib}
\usepackage{natbib} \bibpunct{(}{)}{;}{author-year}{}{,}

%Pour les rajouts
\usepackage{color}
\definecolor{trustcolor}{rgb}{0,0,1}

\usepackage{dsfont}
\usepackage[warn]{textcomp}
\usepackage{adjustbox}
\usepackage{multirow}
\usepackage{subcaption}
\usepackage{graphicx}
\graphicspath{{../figures/}}
\DeclareMathOperator*{\argmin}{\arg\!\min}

\let\tabbeg\tabular
\let\tabend\endtabular
\renewenvironment{tabular}{\begin{adjustbox}{max width=0.95\textwidth}\tabbeg}{\tabend\end{adjustbox}}

\makeatletter

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% LyX specific LaTeX commands.
%% Bold symbol macro for standard LaTeX users
%\newcommand{\boldsymbol}[1]{\mbox{\boldmath $#1$}}

%% Because html converters don't know tabularnewline
\providecommand{\tabularnewline}{\\}
\renewcommand*{\arraystretch}{1.2}

\usepackage{babel}
\makeatother


\begin{document}

\renewcommand{\thefigure}{S\arabic{figure}}
\setcounter{figure}{0}
\renewcommand{\thetable}{S\arabic{table}}
\setcounter{table}{0}
\renewcommand{\theequation}{S\arabic{equation}}
\setcounter{equation}{0}

\section*{Supplementary Materials}

\vspace*{5em}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[h]
	\centerline{\includegraphics[width=0.7\textwidth]{simu-qc-plot}}
	\caption{Quality control plot, as proposed in \cite{prive2020ldpred2}, for the simulations with sample size misspecification. The standard deviations are derived from the summary statistics assuming the same global GWAS sample size for all variants (300,000).}
	\label{fig:qcplot}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{lassosum2-misN}}
	\caption{Multiple results (squared correlation $r^2$ between polygenic score and phenotype) over a grid of parameters ($\lambda$ and $\delta$) for lassosum2 in one of the simulations with misspecified GWAS sample sizes ($n_j$). The different colors indicate whether we use the maximum of $n_j$'s (``maxN'') or the imputed $n_j$'s (``imputeN''), and the normal LD matrix (``normal'') or the one with independent LD blocks (``with\_indep\_ld'').  Missing points represent models that were detected as divergent.}
	\label{fig:lassosum2-misN}
\end{figure}


\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{ldpred2-misN}}
	\caption{Multiple results (squared correlation $r^2$ between polygenic score and phenotype) over a grid of parameters (the SNP heritability $h^2$ and the proportion of causal variants $p$) for LDpred2(-grid) in one of the simulations with misspecified GWAS sample sizes ($n_j$). The different colors indicate whether we use the maximum of $n_j$'s (``maxN'') or the imputed $n_j$'s (``imputeN''), and the normal LD matrix (``normal'') or the one with independent LD blocks (``with\_indep\_ld''). Note that the small $h^2$ values (0.002, 0.02, 0.06) were added and are part of the method we call ``LDpred2-low-h2'' in the main text. Missing points represent models that were detected as divergent.}
	\label{fig:ldpred2-misN}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.8\textwidth]{compare-sd}}
	\caption{Standard deviations from ``true'' genotype calls versus from imputed dosages, colored by INFO scores. [TODO: ALL, MAKE SIDE BY SIDE WITH A/B + DESCRIBE MORE] \label{fig:compare-sd}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.8\textwidth]{compare-beta}}
	\caption{GWAS effect sizes from ``true'' genotype calls versus from imputed dosages, colored by INFO scores. \label{fig:compare-beta}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.8\textwidth]{compare-beta-se}}
	\caption{Standard errors of GWAS effect sizes from ``true'' genotype calls versus from imputed dosages, colored by INFO scores. \label{fig:compare-beta-se}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[p]
	\centerline{\includegraphics[width=0.8\textwidth]{compare-info}}
	\caption{INFO scores reported in the UK Biobank vs.\ INFO scores recomputed from the subset of 362,307 European individuals used in this paper. [TODO: FOR THE VARIANTS IN SIMU?] \label{fig:compare-info}}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.8\textwidth]{compare-mult-imp}}
	\caption{GWAS effect sizes and Z-scores from imputed dosages versus from multiple imputation (MI, 10 draws used), ``corrected'' and colored by INFO scores. This shows only one tenth of the simulated variants, because MI can be computationally intensive. \label{fig:compare-mi}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.65\textwidth]{brca_icogs_hist_info}}
	\caption{Histogram of the imputation INFO scores of the HapMap3 variants from the iCOGS GWAS summary statistics for breast cancer \cite[]{michailidou2013large}. \label{fig:hist-info-brca2}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.65\textwidth]{brca_onco_hist_info}}
	\caption{Histogram of the imputation INFO scores of the HapMap3 variants from the OncoArray GWAS summary statistics for breast cancer \cite[]{michailidou2017association}. [TODO: REEXPORT -- WAS CORRUPTED] \label{fig:hist-info-brca}}
\end{figure}

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.8\textwidth]{hist_bad_pos}}
	\caption{Histogram of the positions of outlier variants from Figure 2B [TODO: VERIF NUMBER AT THE END] in the main text. \label{fig:hist-bad-pos}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{gwas_bad_pc}}
	\caption{Standard deviations inferred from the simulated GWAS summary statistics (\textbf{A:} with no covariate; \textbf{B:} with PC19 from the UK Biobank as covariate) versus the ones inferred from the allele frequencies. Only HapMap3 variants from chromosome 6 are represented.}
	\label{fig:gwas_bad_pc}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{brca_icogs_qc}}
	\caption{Standard deviations inferred from the iCOGS breast cancer GWAS summary statistics versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
	\label{fig:brca_icogs_qc}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{cad_qc}}
	\caption{Standard deviations inferred from the CAD GWAS summary statistics versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
	\label{fig:cad_qc}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{mdd_qc}}
	\caption{Standard deviations inferred from the MDD GWAS summary statistics versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
	\label{fig:mdd_qc}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{prca_qc}}
	\caption{Standard deviations inferred from the PRCA GWAS summary statistics versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random. [TODO: REEXPORT FIGURE]}
	\label{fig:prca_qc}
\end{figure}

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{t1d_affy_qc}}
	\caption{Standard deviations inferred from the T1D (Affymetrix) GWAS summary statistics versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
	\label{fig:t1d_affy_qc}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=0.95\textwidth]{t1d_illu_qc}}
\caption{Standard deviations inferred from the T1D (Illumina) GWAS summary statistics versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
\label{fig:t1d_illu_qc}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.75\textwidth]{res-cad}}
	\caption{Raw partial correlations (for \textit{all} models) for predicting CAD in the UK Biobank from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation.}
	\label{fig:res_cad}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.75\textwidth]{res-mdd}}
	\caption{Raw partial correlations (for \textit{all} models) for predicting MDD in the UK Biobank from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation.}
	\label{fig:res_mdd}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.75\textwidth]{res-prca}}
	\caption{Raw partial correlations (for \textit{all} models) for predicting PRCA in the UK Biobank from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation.}
	\label{fig:res_prca}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{res-t1d}}
	\caption{Raw partial correlations (for \textit{all} models) for predicting T1D in the UK Biobank from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation.}
	\label{fig:res_t1d}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{res-all-brca}}
	\caption{Raw partial correlations (for \textit{all} models and \textit{all} sets) for predicting BRCA in the UK Biobank from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation.}
	\label{fig:res_brca_all}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.9\textwidth]{res-all-cad}}
	\caption{Raw partial correlations (for \textit{all} models and \textit{all} sets) for predicting CAD in the UK Biobank from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation.}
	\label{fig:res_cad_all}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.9\textwidth]{res-all-mdd}}
	\caption{Raw partial correlations (for \textit{all} models and \textit{all} sets) for predicting MDD in the UK Biobank from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation.}
	\label{fig:res_mdd_all}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.9\textwidth]{res-all-prca}}
	\caption{Raw partial correlations (for \textit{all} models and \textit{all} sets) for predicting PRCA in the UK Biobank from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation.}
	\label{fig:res_prca_all}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{res-all-t1d}}
	\caption{Raw partial correlations (for \textit{all} models and \textit{all} sets) for predicting T1D in the UK Biobank from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation.}
	\label{fig:res_t1d_all}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.8\textwidth]{simu_hist_info}}
	\caption{Histogram of the imputation INFO scores of the 40,000 variants from chromosome 22 of the UK Biobank data used in the simulations. \label{fig:hist-info}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\clearpage

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.85\textwidth]{pseudoval}}
	\caption{For one simulation, scores from validation vs.\ scores from pseudo-validation as described in \cite{mak2017polygenic} using either \textbf{A:} correlations (the default in lassosum) or \textbf{B:} p-values when computing local false discovery rates.}
	\label{fig:pseudoval}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{compare-cor}}
	\caption{Comparing correlations between a subset of HapMap3 variants on chromosome 22, computed in three different ways: 1/ from imputed dosages; 2/ from multiple imputation (i.e.\ generating multiple complete datasets sampled according to imputation probabilities, computing correlations, and averaging results); from imputed dosages with the multiplicative correction of $\sqrt{\text{INFO}_j \cdot \text{INFO}_k}$ (for $j \neq k$).}
	\label{fig:compare-cor}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier
\clearpage

\bibliographystyle{natbib}
\bibliography{refs}

\end{document}
