%% LyX 1.3 created this file.  For more info, see http://www.lyx.org/.
%% Do not edit unless you really know what you are doing.
\documentclass[english, 12pt]{article}
\usepackage{times}
%\usepackage{algorithm2e}
\usepackage{url}
\usepackage{bbm}
\usepackage[T1]{fontenc}
\usepackage[latin1]{inputenc}
\usepackage{geometry}
\geometry{verbose,letterpaper,tmargin=2cm,bmargin=2cm,lmargin=1.5cm,rmargin=1.5cm}
\usepackage{rotating}
\usepackage{color}
\usepackage{graphicx}
\usepackage{amsmath, amsthm, amssymb}
\usepackage{setspace}
\usepackage{lineno}
\usepackage{hyperref}
\usepackage{bbm}
\usepackage{makecell}
\usepackage{placeins}
\usepackage{subcaption}

%\renewcommand{\arraystretch}{1.8}

%\linenumbers
%\doublespacing
\onehalfspacing
%\usepackage[authoryear]{natbib}
\usepackage{natbib} %\bibpunct{(}{)}{;}{author-year}{}{,}

%Pour les rajouts
\usepackage{color}
\definecolor{trustcolor}{rgb}{0,0,1}

\usepackage{dsfont}
\usepackage[warn]{textcomp}
\usepackage{adjustbox}
\usepackage{multirow}
\usepackage{subcaption}
\usepackage{graphicx}
\graphicspath{{../figures/}}
\DeclareMathOperator*{\argmin}{\arg\!\min}

\let\tabbeg\tabular
\let\tabend\endtabular
\renewenvironment{tabular}{\begin{adjustbox}{max width=0.95\textwidth}\tabbeg}{\tabend\end{adjustbox}}

\makeatletter

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% LyX specific LaTeX commands.
%% Bold symbol macro for standard LaTeX users
%\newcommand{\boldsymbol}[1]{\mbox{\boldmath $#1$}}

%% Because html converters don't know tabularnewline
\providecommand{\tabularnewline}{\\}
\renewcommand*{\arraystretch}{1.2}

\usepackage{babel}
\makeatother


\begin{document}

\renewcommand{\thefigure}{S\arabic{figure}}
\setcounter{figure}{0}
\renewcommand{\thetable}{S\arabic{table}}
\setcounter{table}{0}
\renewcommand{\theequation}{S\arabic{equation}}
\setcounter{equation}{0}

\section*{Supplementary Materials}

\vspace*{5em}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[h]
	\centerline{\includegraphics[width=0.7\textwidth]{simu-qc-plot}}
	\caption{Quality control plot, as proposed in \cite{prive2020ldpred2}, for the simulations with sample size misspecification. The standard deviations are derived from the summary statistics assuming the same global GWAS sample size for all variants (300,000).}
	\label{fig:qcplot}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{lassosum2-misN}}
	\caption{Multiple results (squared correlation $r^2$ between polygenic score and phenotype) over a grid of parameters ($\lambda$ and $\delta$) for lassosum2 in one of the simulations with misspecified GWAS sample sizes ($n_j$). The different colors indicate whether we use the maximum of $n_j$'s (``maxN'') or the imputed $n_j$'s (``imputeN''), and the normal LD matrix (``normal'') or the one with independent LD blocks (``with LD blocks'').  Missing points represent models that were detected as divergent.}
	\label{fig:lassosum2-misN}
\end{figure}


\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{ldpred2-misN}}
	\caption{Multiple results (squared correlation $r^2$ between polygenic score and phenotype) over a grid of parameters (the SNP heritability $h^2$ and the proportion of causal variants $p$) for LDpred2(-grid) in one of the simulations with misspecified GWAS sample sizes ($n_j$). The different colors indicate whether we use the maximum of $n_j$'s (``maxN'') or the imputed $n_j$'s (``imputeN''), and the normal LD matrix (``normal'') or the one with independent LD blocks (``with LD blocks''). Note that the small $h^2$ values (0.002, 0.02, 0.06) were added and are part of the method we call ``LDpred2-low-h2'' in the main text. Missing points represent models that were detected as divergent.}
	\label{fig:ldpred2-misN}
\end{figure}

\begin{figure}[!h]
	\centerline{\includegraphics[width=0.95\textwidth]{simu-misN-smallh2}}
	\caption{Results for the simulations with sample size misspecification and a heritability of 4\%, averaged over 10 simulations for each scenario. Reported 95\% confidence intervals are computed from 10,000 non-parametric bootstrap replicates of the mean. The GWAS sample size is either ``true'' when providing the true per-variant sample size, ``max'' when providing instead the maximum sample size as a unique value to be used for all variants, ``imputed'', or ``any'' when the method does not use this information (the case for C+T). Red bars correspond to using the LD with independent blocks, which is a requirement in lassosum and PRS-CS.}
	\label{fig:simu-misN-smallh2}
\end{figure}

\begin{figure}[!h]
	\centerline{\includegraphics[width=\textwidth]{simu-misN-shrunk}}
	\caption{Results for the simulations with sample size misspecification and a heritability of 20\% (bottom panels) or 4\% (top panels), averaged over 10 simulations for each scenario. Reported 95\% confidence intervals are computed from 10,000 non-parametric bootstrap replicates of the mean. The GWAS sample size is either ``true'' when providing the true per-variant sample size, ``max'' when providing instead the maximum sample size as a unique value to be used for all variants, or ``imputed''. Red bars correspond to using the LD with independent blocks. Bars filled in blue correspond to using the shrunk LD matrix computed from GCTB \cite[]{lloyd2019improved}, otherwise the windowed LD matrix from LDpred2 is used.}
	\label{fig:simu-misN-shrunk}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.8\textwidth]{simu_hist_info}}
	\caption{Histogram of the imputation INFO scores of the 40,000 variants from chromosome 22 of the UK Biobank data used in the simulations. \label{fig:hist-info}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{compare-sd}}
	\caption{\textbf{A:} Raw standard deviations (SDs) from imputed dosages or \textbf{B:} corrected SDs from imputed dosages (dividing them by $\sqrt{\text{INFO}}$) versus SDs from ``true'' genotype calls, colored by INFO scores. \label{fig:compare-sd}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{compare-beta}}
	\caption{GWAS effect sizes ($\hat{\gamma}$) from ``true'' genotype calls versus either \textbf{A:} raw $\hat{\gamma}$ from imputed dosages or \textbf{B:} corrected $\hat{\gamma}$ from imputed dosages (multiplying them by $\sqrt{\text{INFO}}$), colored by INFO scores. \label{fig:compare-beta}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=0.95\textwidth]{compare-beta-se}}
\caption{Standard errors of GWAS effect sizes (SEs) from ``true'' genotype calls versus either \textbf{A:} raw SEs from imputed dosages or \textbf{B:} corrected SEs from imputed dosages (multiplying them by $\sqrt{\text{INFO}}$), colored by INFO scores. \label{fig:compare-beta-se}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=0.95\textwidth]{compare-mult-imp}}
\caption{(Over)corrected GWAS \textbf{A:} Z-scores and \textbf{B:} effect sizes from imputed dosages (multiplying them by $\text{INFO}$) versus the ones obtained from multiple imputation (MI, 20 draws used, \cite{Palmer2016}), colored by INFO scores. This shows only one tenth of the simulated variants, because MI can be computationally intensive. \label{fig:compare-mi}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.75\textwidth]{compare-info}}
	\caption{INFO scores reported in the UK Biobank for variants used in simulations versus INFO scores recomputed from the subset of 362,307 European individuals used in this paper. \label{fig:compare-info}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.98\textwidth]{compare-all-maf-info-chr22}}
	\caption{Minor allele frequencies (MAF) and INFO scores in the UK Biobank for 50,000 variants on chromosome 22 (chosen at random). These are either reported (in MFI files from the UK Biobank), recomputed from the whole data or from the subset of 362,307 European individuals used in this paper. MAF are represented on a log scale. \label{fig:compare-maf-info}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.9\textwidth]{simu-rounded}}
	\caption{Results for the simulations with summary statistics (effect sizes and their standard errors) possibly rounded to two significant digits, averaged over 10 simulations for each scenario. Reported 95\% confidence intervals are computed from 10,000 non-parametric bootstrap replicates of the mean. Red bars correspond to using the LD with independent blocks (Methods).}
	\label{fig:simu-rounded}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.65\textwidth]{brca_icogs_hist_info}}
	\caption{Histogram of the imputation INFO scores of the HapMap3 variants from the iCOGS GWAS summary statistics for breast cancer \cite[]{michailidou2013large}. \label{fig:hist-info-brca2}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.65\textwidth]{brca_onco_hist_info}}
	\caption{Histogram of the imputation INFO scores of the HapMap3 variants from the OncoArray GWAS summary statistics for breast cancer \cite[]{michailidou2017association}. \label{fig:hist-info-brca}}
\end{figure}

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.75\textwidth]{hist_bad_pos}}
	\caption{Histogram of the positions of outlier variants from Figure 5B in the main text.%TODO: VERIF NUMBER AT THE END
	\label{fig:hist-bad-pos}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{gwas_bad_pc}}
	\caption{Standard deviations inferred from the simulated GWAS summary statistics (\textbf{A:} with no covariate; \textbf{B:} with PC19 from the UK Biobank as covariate) versus the ones inferred from the allele frequencies. Only HapMap3 variants from chromosome 6 are represented.}
	\label{fig:gwas_bad_pc}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{brca_icogs_qc}}
	\caption{Standard deviations inferred from the iCOGS breast cancer GWAS summary statistics (\textbf{A:} Raw or \textbf{B:} dividing them by $\sqrt{\text{INFO}}$) versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
	\label{fig:brca_icogs_qc}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{cad_qc}}
	\caption{Standard deviations inferred from the CAD GWAS summary statistics (\textbf{A:} Raw or \textbf{B:} dividing them by $\sqrt{\text{INFO}}$) versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
	\label{fig:cad_qc}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{mdd_qc}}
	\caption{Standard deviations inferred from the MDD GWAS summary statistics (\textbf{A:} Raw or \textbf{B:} dividing them by $\sqrt{\text{INFO}}$) versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
	\label{fig:mdd_qc}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{prca_qc}}
	\caption{Standard deviations inferred from the PRCA GWAS summary statistics (\textbf{A:} Raw or \textbf{B:} dividing them by $\sqrt{\text{INFO}}$) versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
	\label{fig:prca_qc}
\end{figure}

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{t1d_affy_qc}}
	\caption{Standard deviations inferred from the T1D (Affymetrix) GWAS summary statistics (\textbf{A:} Raw or \textbf{B:} dividing them by $\sqrt{\text{INFO}}$) versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
	\label{fig:t1d_affy_qc}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=0.95\textwidth]{t1d_illu_qc}}
\caption{Standard deviations inferred from the T1D (Illumina) GWAS summary statistics (\textbf{A:} Raw or \textbf{B:} dividing them by $\sqrt{\text{INFO}}$) versus the ones inferred from the reported GWAS allele frequencies. Only 100,000 variants are represented, at random.}
\label{fig:t1d_illu_qc}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.7\textwidth]{res-brca_icogs}}
	\caption{Variance explained of BRCA in the UK Biobank by PGS derived from external summary statistics (iCOGS). These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation; these values are then squared.}
	\label{fig:res_brca_icogs}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.7\textwidth]{res-brca_onco}}
	\caption{Variance explained of BRCA in the UK Biobank by PGS derived from external summary statistics (OncoArray). These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation; these values are then squared.}
	\label{fig:res_brca_onco}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.7\textwidth]{res-mdd}}
	\caption{Variance explained of MDD in the UK Biobank by PGS derived from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation; these values are then squared.}
	\label{fig:res_mdd}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.7\textwidth]{res-prca}}
	\caption{Variance explained of PRCA in the UK Biobank by PGS derived from external summary statistics. These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation; these values are then squared.}
	\label{fig:res_prca}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.7\textwidth]{res-t1d_illu}}
	\caption{Variance explained of T1D in the UK Biobank by PGS derived from external summary statistics (Illumina). These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation; these values are then squared.}
	\label{fig:res_t1d_illu}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.7\textwidth]{res-t1d_affy}}
	\caption{Variance explained of T1D in the UK Biobank by PGS derived from external summary statistics (Affymetrix). These are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation; these values are then squared (while keeping the sign).}
	\label{fig:res_t1d_aff}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\clearpage

\FloatBarrier

\begin{figure}[p]
\centerline{\includegraphics[width=0.8\textwidth]{vitaminD_qc}}
\caption{Standard deviations inferred from the vitamin D GWAS summary statistics versus the ones inferred from the allele frequencies of the LD reference. Only 100,000 variants are represented, at random.}
\label{fig:vitaminD_qc}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=0.95\textwidth]{hist_N_vitaminD}}
\caption{Histogram of the per-variant sample sizes in the vitamin D GWAS summary statistics. The vertical red line corresponds to 70\% of the maximum sample size, the threshold used in ``qc2''. \label{fig:hist-N-vitaminD}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{res-BBJ}}
	\caption{Results for PGS derived from four Biobank Japan GWAS summary statistics and using three different LD references. Partial correlations are computed using function \texttt{pcor} of R package bigstatsr where 95\% confidence intervals are obtained through Fisher's Z-transformation, then all values are squared to report the phenotypic variance explained by PGS.}
	\label{fig:bbj}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.85\textwidth]{pseudoval}}
	\caption{For one simulation, scores from validation versus scores from pseudo-validation as described in \cite{mak2017polygenic} using either \textbf{A:} correlations (the default in lassosum) or \textbf{B:} p-values, when computing local false discovery rates.}
	\label{fig:pseudoval}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{compare-cor}}
	\caption{Comparing pairwise correlations between a subset of HapMap3 variants on chromosome 1 with an INFO score lower than 0.85, computed in three different ways: 1/ (left y-axis) from imputed dosages using 10,000 individuals from the UK Biobank (UKBB) data; 2/ (right y-axis) from multiple imputation (i.e.\ generating multiple complete datasets sampled according to imputation probabilities, computing correlations, and averaging results) also using UKBB; 3/ (common x-axis) from 190 individuals from the 1000 Genomes data (GBR and CEU). Each point, representing the correlation between two variants, is colored by the geometric mean of the INFO scores of these two variants.}
	\label{fig:compare-cor}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier
\clearpage

\bibliographystyle{natbib}
\bibliography{refs}

\end{document}
